### ISSUE TITLE
[REVIEW]: Ripserer.jl: flexible and efficient persistent homology computation in Julia

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2614]
**Submitting author:** @mtsch (<a href="http://orcid.org/0000-0003-0734-2719">Matija Čufar</a>)
**Repository:** <a href="https://github.com/mtsch/Ripserer.jl" target ="_blank">https://github.com/mtsch/Ripserer.jl</a>
**Version:** v0.15.2
**Editor:** @VivianePons
**Reviewer:** @juliohm, @sauln
**Archive:** <a href="https://doi.org/10.5281/zenodo.4081109" target="_blank">10.5281/zenodo.4081109</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9/status.svg)](https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9"><img src="https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9/status.svg)](https://joss.theoj.org/papers/0c8b6abead759ba068ee178fedc998a9)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/mtsch/Ripserer.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@juliohm & @sauln, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @VivianePons know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mtsch/Ripserer.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mtsch) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mtsch/Ripserer.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mtsch) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @juliohm, @sauln it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-64185-0_11 is OK
- 10.1137/15M1031151 is OK
- 10.1038/ncomms15396 is OK
- 10.1145/3350755.3400244 is OK
- 10.4230/LIPIcs.SoCG.2016.71 is OK
- 10.21105/joss.00925 is OK
- 10.1137/141000671 is OK

MISSING DOIs

- 10.1090/conm/453/08802 may be a valid DOI for title: Persistent homology-a survey

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02614/joss.02614/10.21105.joss.02614.pdf)

**Comment by mtsch** [comment]
I fixed the missing DOI.

@whedon generate pdf

**Comment by mtsch** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02614/joss.02614/10.21105.joss.02614.pdf)

**Comment by juliohm** [comment]
I love to review submissions such as this, really well-written, to the point, and a great piece of software! Thanks @mtsch for this amazing work. I only have one minor comment that I already shared as a GitHub issue yesterday about the related Julia packages you recently added to the docs: https://mtsch.github.io/Ripserer.jl/dev/related-work/

Do you think that these should also be listed in the JOSS paper or at least cited in the list of references? I noticed that you cited the major alternatives from other languages, but perhaps it would be a good idea to list related work in Julia.

**Comment by mtsch** [comment]
Thanks for the kind words!

I added a short paragraph listing Julia packages. I also added citation for Julia and a reference to a Ripser fork I forgot to include.

**Comment by mtsch** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02614/joss.02614/10.21105.joss.02614.pdf)

**Comment by juliohm** [comment]
I am happy to approve the submission. Everything is checked on my side :100:

**Comment by VivianePons** [comment]
Thank you very much @juliohm for reviewing so efficiently! @sauln let me know when you are done and do not hesitate to ask questions / make comments to the author, also you can take your time ;) The other reviewer was particularly quick!

**Comment by VivianePons** [comment]
Hi @sauln I haven't heard of you in a while and nothing has been udptaed on the review, are you still on board?

**Comment by sauln** [comment]
Hi Sorry @VivianePons  It's still on my radar but have been swamped the last few weeks. I'll try to either finish it up this weekend, or bow out.

**Comment by VivianePons** [comment]
If you need some time, it's ok. I just need to know we are still on track

**Comment by VivianePons** [comment]
Hi @sauln I see that you haven't finished it up as you were planning (which is ok) but I'm hoping that you haven't bowed out. If you have, please let me know!

**Comment by sauln** [comment]
@VivianePons, I have completed my review.  As @juliohm expressed, this project was a pleasure to review. I am only sorry it took me so long to complete it.  The documentation is best in class, the library interface is remarkably clean and intuitive, and the testing appears more exhaustive than any other TDA library I've seen. 

I am happy to accept this submission to JOSS 🥇

**Comment by mtsch** [comment]
@juliohm and @sauln thanks again for taking the time to review!

**Comment by VivianePons** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02614/joss.02614/10.21105.joss.02614.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02614/joss.02614/10.21105.joss.02614.pdf) :page_facing_up: :point_left:

**Comment by VivianePons** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-64185-0_11 is OK
- 10.1137/15M1031151 is OK
- 10.1090/conm/453/08802 is OK
- 10.1038/ncomms15396 is OK
- 10.1145/3350755.3400244 is OK
- 10.4230/LIPIcs.SoCG.2016.71 is OK
- 10.21105/joss.00925 is OK
- 10.1137/141000671 is OK
- 10.3390/a13010019 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by VivianePons** [comment]
@mtsch I see you give many arXiv references in the bibliography. the arXiv number is always given but the actual link is present only in some ref. Could you put the link on all of them, that makes it easier for the reader to follow the ref!

Once this is done, I need you to make an archive (on Zenodo, or figshare, or other) and give the doi.

Thank you very much!

**Comment by mtsch** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02614/joss.02614/10.21105.joss.02614.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02614/joss.02614/10.21105.joss.02614.pdf) :page_facing_up: :point_left:

**Comment by mtsch** [comment]
@VivianePons all references have links now. I made an archive on Zenodo. [Here](https://zenodo.org/record/4081109) is a link.

**Comment by VivianePons** [comment]
Thanks, could you update the meta information of the Zenodo archive so that it has the same title as the paper and complete author name, thank you!

**Comment by mtsch** [comment]
Done.

**Comment by VivianePons** [comment]
@whedon set v0.15.2 as version

**Comment by whedon** [comment]
OK. v0.15.2 is the version.

**Comment by VivianePons** [comment]
@whedon set 10.5281/zenodo.4081109 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4081109" target="_blank">10.5281/zenodo.4081109</a> is the archive.

**Comment by VivianePons** [comment]
Thank you all for the great work! The paper and software have received many praises, it is pleasure to accept it into JOSS, thank you @mtsch for the work and thank you @juliohm and @sauln for the reviews :)

**Comment by VivianePons** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1826

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1826, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-64185-0_11 is OK
- 10.1137/15M1031151 is OK
- 10.1090/conm/453/08802 is OK
- 10.1038/ncomms15396 is OK
- 10.1145/3350755.3400244 is OK
- 10.4230/LIPIcs.SoCG.2016.71 is OK
- 10.21105/joss.00925 is OK
- 10.1137/141000671 is OK
- 10.3390/a13010019 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mtsch** [comment]
@VivianePons thank you for editing! Submitting to this paper was an awesome experience!

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1318213473806012417) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1827
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02614](https://doi.org/10.21105/joss.02614)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congrats @mtsch on your article's publication in JOSS!

Many thanks to @juliohm and @sauln for reviewing this, and @VivianePons for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02614/status.svg)](https://doi.org/10.21105/joss.02614)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02614">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02614/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02614/status.svg
   :target: https://doi.org/10.21105/joss.02614
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02614/status.svg)](https://doi.org/10.21105/joss.02614)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
